backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/highlighter/impl/LuceneUnifiedSearchHighlighter.java

```diff
@@ -139,37 +143,110 @@ else if ( BoundaryScannerType.SENTENCE.equals( LuceneUnifiedSearchHighlighter.th
 
 		@Override
 		public List<String> highlight(int doc) throws IOException {
-			Map<String, String[]> highlights = highlighter.highlightFields(
-					fieldsIn, query, new int[] { leafReaderContext.docBase + doc }, maxPassagesIn );
-
-			List<String> result = new ArrayList<>();
-			for ( String highlight : highlights.get( fieldsIn[0] ) ) {
-				if ( highlight != null ) {
-					result.add( highlight );
+			List<TextFragment> highlights = highlighter.highlightField( fieldsIn, query, leafReaderContext.docBase + doc, maxPassagesIn );
+			if ( highlights == null ) {
+				return Collections.emptyList();
+			}
+			else {
+				if ( Boolean.TRUE.equals( LuceneUnifiedSearchHighlighter.this.orderByScore ) ) {
+					highlights.sort( Comparator.comparingDouble( TextFragment::score ).reversed() );
 				}
+				List<String> result = new ArrayList<>( highlights.size() );
+				for ( TextFragment highlight : highlights ) {
+					result.add( highlight.highlightedText() );
+				}
+				return result;
 			}
-			return result;
+		}
+	}
+
+	private static class MultiValueUnifiedHighlighter extends UnifiedHighlighter {
+
+		public MultiValueUnifiedHighlighter(IndexSearcher indexSearcher, Analyzer indexAnalyzer) {
+			super( indexSearcher, indexAnalyzer );
+		}
+
+		@SuppressWarnings( "unchecked" )
+		public List<TextFragment> highlightField(String[] fieldIn, Query query, int doc, int[] maxPassagesIn) throws IOException {
+			assert fieldIn.length == 1;
+			return (List<TextFragment>) highlightFieldsAsObjects( fieldIn, query, new int[] { doc }, maxPassagesIn ).get( fieldIn[0] )[0];
 		}
 	}
 
 	// `DefaultPassageFormatter` uses a string builder to create a resulting highlighted string.
-	// Elasticsearch uses their own formatter that creates an array of snippets instead.
-	// Hence, the results won't match if more than one snipped is needed to provide the results.
-	private static class PassageFormatterWithEncoder extends DefaultPassageFormatter {
+	// This is a copy of a `DefaultPassageFormatter` that instead of doing string concatenation is putting each passage
+	// as a new element:
+	private static class PassageFormatterWithEncoder extends PassageFormatter {
```

Comment:
Looks like [coverage for this class is not perfect](https://sonarcloud.io/code?id=org.hibernate.search%3Ahibernate-search-parent&pullRequest=3496&selected=org.hibernate.search%3Ahibernate-search-parent%3Abackend%2Flucene%2Fsrc%2Fmain%2Fjava%2Forg%2Fhibernate%2Fsearch%2Fbackend%2Flucene%2Fsearch%2Fhighlighter%2Fimpl%2FLuceneUnifiedSearchHighlighter.java&line=179)... Unless I'm wrong it's a rather critical piece of code, so do you think you could add a few unit tests? Or alternatively an integration tests that involves overlapping terms, whatever you prefer.

Reply:
yeah 🥲😃  will try with IT first to keep this class private.

Reply:
Hmmm, so I've tried to come up with a case where there'll be overlaps. From what I can see, it probably won't happen since a unified highlighter cannot "highlight phrases"... So now I'm not sure if we really need to keep that loop ... I've also looked at the Lucene test for this passage formatter and there's no case for overlapping 🙈. I suppose we can create a test for a formatter and build passages ourselves so that they have overlaps, but I'm not sure if we should if that's not going to happen... 
I've added a line to the docs mentioning that phrase highlighting is only possible with FVH.

Reply:
Are you sure it's not possible to get overlaps with non-phrase highlighting?

I know there are analyzers that will turn `wi-fi` into three terms: `wi`, `fi` and `wifi`. With such an analyzer, a query such as `OR('wi', 'wifi')` would potentially match both the whole original term (`wi-fi`) and part of it (`wi`), so there would be overlap... ?

Anyway, I'd indeed recommend just a unit test, since it looks rather complicated to set up in an integration test.